compiler/src/dotty/tools/dotc/semanticdb/SemanticSymbolBuilder.scala

```diff
@@ -0,0 +1,124 @@
+  private val locals = mutable.HashMap[Symbol, Int]()
+
+  /** The local symbol(s) starting at given offset */
+  private val symsAtOffset = new mutable.HashMap[Int, Set[Symbol]]():
+    override def default(key: Int) = Set[Symbol]()
+
+  extension (sym: Symbol)
+    /** The semanticdb name of the given symbol */
+    def symbolName(using Context): String =
+      val b = StringBuilder(20)
+      addSymName(b, sym)
+      b.toString
+
+    def funParamSymbol(using Context): Name => String =
+      if sym.isGlobal then
+        val funSymbol = sym.symbolName
+        name => s"$funSymbol($name)"
+      else
+        name => locals.keys.find(local => local.isTerm && local.owner == sym && local.name == name)
+                      .fold("<?>")(Symbols.LocalPrefix + _)
+
+  /** Add semanticdb name of the given symbol to string builder */
+  private def addSymName(b: StringBuilder, sym: Symbol)(using Context): Unit =
+
+    def addName(name: Name) =
+      val str = name.toString.unescapeUnicode
+      if str.isJavaIdent then b append str
+      else b append '`' append str append '`'
+
+    def addOwner(owner: Symbol): Unit =
+      if !owner.isRoot then addSymName(b, owner)
+
+    def addOverloadIdx(sym: Symbol): Unit =
+      val decls =
+        val decls0 = sym.owner.info.decls.lookupAll(sym.name)
+        if sym.owner.isAllOf(JavaModule) then
+          decls0 ++ sym.owner.companionClass.info.decls.lookupAll(sym.name)
+        else
+          decls0
+      end decls
+      val alts = decls.filter(_.isOneOf(Method | Mutable)).toList.reverse
+      def find(filter: Symbol => Boolean) = alts match
+        case notSym :: rest if !filter(notSym) =>
+          val idx = rest.indexWhere(filter).ensuring(_ >= 0)
+          b.append('+').append(idx + 1)
+        case _ =>
+      end find
+      val sig = sym.signature
+      find(_.signature == sig)
+
+    def addDescriptor(sym: Symbol): Unit =
+      if sym.is(ModuleClass) then
+        addDescriptor(sym.sourceModule)
+      else if sym.is(TypeParam) then
+        b.append('['); addName(sym.name); b.append(']')
+      else if sym.is(Param) then
+        b.append('('); addName(sym.name); b.append(')')
+      else if sym.isRoot then
+        b.append(Symbols.RootPackage)
+      else if sym.isEmptyPackage then
+        b.append(Symbols.EmptyPackage)
+      else if (sym.isScala2PackageObject) then
+        b.append(Symbols.PackageObjectDescriptor)
+      else
+        addName(sym.name)
+        if sym.is(Package) then b.append('/')
+        else if sym.isType || sym.isAllOf(JavaModule) then b.append('#')
+        else if sym.isOneOf(Method | Mutable)
+        && (!sym.is(StableRealizable) || sym.isConstructor) then
+          b.append('('); addOverloadIdx(sym); b.append(").")
+        else b.append('.')
+
+    /** The index of local symbol `sym`. Symbols with the same name and
+     *  the same starting position have the same index.
+     */
+    def localIdx(sym: Symbol)(using Context): Int =
```

Comment:
Sry for confustion, I reverted https://github.com/lampepfl/dotty/pull/12885/commits/58fd3d581bba5a5190d7a0f6a1f5ab538b1b8612 because it's not already needed 👍